nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol

```diff
@@ -1216,6 +1216,11 @@ contract StakingEscrow is Issuer {
         bytes32 memoryAddress = delegateGetData(_target, "stakerInfo(address)", 1, _staker, 0);
         assembly {
             result := memoryAddress
+            // copy data to the right position after the array pointer place
+            // measureWork
+            mstore(add(memoryAddress, 0x140), mload(add(memoryAddress, 0x100)))
```

Comment:
Is this direct memory addressing the result of previous version not having those two fields? E.g. is it how we will handle similar upgrades in future?

Reply:
It's independent from previous version, issue that structure has array pointers between fields. And function `stakerInfo` returns without array pointers but without gap. To correct translate from returned value to `StakerInfo` struct need to move tail of memory. In future need to add copying for each new slot of memory

Reply:
Still not understanding fully - So this is specific to the upgrade issue with different struct layouts?  If we deploy on another network, with this as the initially deployed contract, then this code will have no effect (rather what effect will it have)?

Reply:
I'm afraid I too don't understand. Why is `measureWork` (L1220) relevant here?

Reply:
Oh, is it because `pastDowntime` and `subStakes` are arrays? and these arrays in the struct have a different memory structure in storage and when returned by `stakerInfo`?

Reply:
Every layout (in every contract) such as 
```
uint256 value1;
uint256[] array1;
uint256 value2;
```
needs memory copying in this type of function.
Mostly we try to have this type of layout:
```
uint256 value1;
uint256 value2;
uint256[] array1;
```
because no need to copy memory
But backward compatibility requires add new field in the end of the structure.
So in new network/deployment this contract still requires  memory copying but we have opportunity to change layout before deploy if we want and remove unnesessary code

Reply:
@cygnusv 
Storage layout (https://solidity.readthedocs.io/en/latest/miscellaneous.html?highlight=layout#layout-of-state-variables-in-storage)
`uint256 value;` -> 1st slot, `memoryPointer`
`uint16 confirmedPeriod1;`  and `uint16 confirmedPeriod2;` -> 2nd slot , `memoryPointer` + 0x20
`bool reStake;` and `uint16 lockReStakeUntilPeriod;` -> 3rd slot, `memoryPointer` + 0x40
`address worker;` -> 4th slot, `memoryPointer` + 0x60
`uint16 workerStartPeriod;` and `uint16 lastActivePeriod;` -> 5th slot, `memoryPointer` + 0x80
pointer to `Downtime[] pastDowntime;` in 6th slot, `memoryPointer` + 0x100
pointer to `SubStakeInfo[] subStakes;` in 7th slot, `memoryPointer` + 0x120
`bool measureWork;` in 8th slot, `memoryPointer` + 0x140
`uint256 completedWork;` in 9th slot, `memoryPointer` + 0x160

Can't say for sure why only two values in one slot for small types

`stakerInfo` returns without array pointers and `measureWork` `completedWork` are shifted to the left. To assign returned value to structure need to shift them to the right

Reply:
OK, so this part of the code address the problem of memory copying because of current structure that contains arrays in the middle of the struct. Your point is that in the future we have the chance to deploy a Testnet V2 with a simpler memory structure. Is that correct?

Reply:
Yes, it's correct. I prefer not to change this, but we can

Reply:
Yeah, I'd certainly prefer to start mainnet without those unobvious pointers, which means that we need to have some testnet without those pointers.

But good to test those backward compatibility measures in practice!

Reply:
> Yeah, I'd certainly prefer to start mainnet without those unobvious pointers

> But good to test those backward compatibility measures in practice!

I agree!

Reply:
Does it make sense then, instead of doing this fix, have a hard fork with the correct structure of `stakerInfo`?

Reply:
This means we'd also need to redeploy the token, start with 0 stakes, redistribute, etc.  considering our current standing - I think this fix is still the best course of action for *this* deployment set.

Reply:
One of the best additions we can make it some tooling to help mitigate this class of problem in the future.  How can we prevent this in CI?  (Perhaps another discussion)

Reply:
I'm for having this code maybe even in production, because it shows narrow places and we will meet with those type of upgrading on mainnet.
Other idea - do hardfork test (before mainnet), not just redeploy but also break contract and copy information from old to new one

Reply:
Well, we don't need to be attached for this deployment set. If we deploy this new contract then at some point we'll still have to do hard fork if we want to deploy in mainnet with the correct structure from the start.

My point is that now is the moment to do unorthodox changes, not just before mainnet.

Reply:
@cygnusv I totally agree with that. Nevertheless, it is very valuable to test, how we will do upgrades in similar cases.
I think, incentivized testnet is a good candidate for a "clean" deployment without this patch